### ISSUE TITLE
Issue with Google Analytics 4

### ISSUE THREAD

**Issue by WoMec** [rainlab/googleanalytics-plugin#101]
Hi, it seems that plugin doesn't work with "new" Google Analytics 4 ... If I create a new property you GA, I'm no longer available to create a new view and get tracking ID.  Can you please check, if your plugin works with a new GA4 properties? 
Thanks

**Comment by LukeTowers** [comment]
What do you mean you can't create a new view and get a tracking ID? That's not handled by this plugin AFAIK, it's handled by Google Analytics itself.

**Comment by WoMec** [comment]
Sure, I know. To get plugin working, you have to setup GA to get private key (json file), profile ID and Tracking ID. I have your plugin successfully running in another website with "old" GA. But now, I created new property in GA, and there is no longer "VIEW" section, so I don't know the Profile ID. Can you please check, if I'm doing something wrong, or if there is realy some change in the new Google Analytics 4 ? 
Thanks.

**Comment by ayumi-cloud** [comment]
@LukeTowers Please can you open Google Analytics! There is no such thing now as `Profile ID` it has been scraped and the only things I can find is `PROPERTY ID` under the **Admin > Property Settings** 

Likewise, Google Analytics is removed and only Google Tag Manager is now showing, under **Admin > Setup Assistant > Tag installation** 

You then get a `MEASUREMENT ID`

Also there is a new `Streams API`, under **Admin > Setup Assistant > Tag installation** 

You then get a `STREAM ID`

Basically, this plugin is dead in the water and no longer works at all, because it doesn't connect up with Google Tag Manager or Streams.

**Comment by ayumi-cloud** [comment]
Ok done some research, this plugin needs to be re-coded to use this future api:

Measurement Protocol (Google Analytics 4), see here: https://developers.google.com/analytics/devguides/collection/protocol/ga4

> Warning: This is an alpha API and subject to change. You may encounter breaking changes while it is in alpha. Code using this API should not be pushed to production. See limitations for issues that will be address before a general availability launch.

**Comment by ayumi-cloud** [comment]
I should also mention, due to GDPR laws Google Analytics 4 IP addresses are automatically anonymized (limiting the data coming to October).

**Comment by Eoler** [comment]
Please read this: https://www.nwsdigital.com/Blog/Should-GA-4-Be-Your-Only-Analytics
and this: https://support.google.com/analytics/thread/79367900?hl=en

I suggest using both Universal (web+app) and GA4 (if you really fear of missing out alpha service).

**Comment by ayumi-cloud** [comment]
### v3 to v4

I wouldn't continue to use `Universal` it's gonna get turned off. Better to move forward with v4.

Two reasons:

1. There are way more features in v4 with `MEASUREMENT` and `STREAMS` plus many other things.

2. The reason why Google is moving from v3 to v4 is because of GDPR issues! As some of our websites are hosted in Europe it's better to not get in hot water and just be compliant with the laws. Google clearly says in it's legal documents that it's down to it's users and not their fault with the data. Though this was always bs and they secretly updated their code to now pass gdpr laws. Switching to v4 and removing v3 makes our websites pass legally automatically now - so best for the end-users privacy as well.

### GA vs other Analytics

As for using Google Analytics and not another Analytics Platform, our clients use GA for SEO purposes, it has some SEO features that other Analytics Platforms don't have.

We also update our clients GA with custom dashboards, that give our clients hundreds of new and powerful widgets. Takes a few seconds to install in GA.

### October Dashboard

We have created our own Analytics Platforms inside October built directly into a firewall.

We were using this plugin and the GA Extension plugin to display SEO data, such as client `goals` etc.

---

That's how we were using this plugin ...

**Comment by ayumi-cloud** [comment]
@LukeTowers 

> When exactly are they going to "turn off" universal analytics?

From what I've seen so far doing some research, it seems like they are going to be turning off the following:

- Universal Analytics `Tracking ID` (which begins with "UA-") will be turned off and replaced with `Measurement ID` (which begins with "G-").

- `analytics.js` will be turned off and replaced with `gtag.js`

- `PROPERTY ID` seems to be removed and I can't find a replacement? Seems like it's changed to either `Measurement ID` or `STREAM ID`

- Measurement Protocol API has changed from v3 to v4. However, v4 is in `Aplha`, see here: , see here: https://developers.google.com/analytics/devguides/collection/protocol/ga4

- AMP-HTML `<amp-analytics>` now doesn't work at all and after an issue we opened Google added this pr: https://github.com/ampproject/amphtml/pull/31095/files

- Google's most popular Analytics extension / plugin for Google Analytics called: `AutoTrack`

![image](https://user-images.githubusercontent.com/57409060/99252786-6c318980-2807-11eb-87a5-fcb7cf9bf142.png)

They have turned off, see here: https://github.com/googleanalytics/autotrack/commit/646e079dbdad6dfb698d1fb4b5047d7edfcb07ef

- GDPR, in v3 the data didn't pass and in v4 various changes were automatically set.

Probably lots more things.

Universal Analytics versus Google Analytics 4 data: https://support.google.com/analytics/answer/9964640?hl=en

Google Analytics Release Logs: https://support.google.com/analytics/answer/9164320?hl=en&ref_topic=9143232

**Comment by ayumi-cloud** [comment]
I forgot to mention:

`Google Analytics Firebase` see here: https://firebase.google.com/docs/analytics

Not sure what's happening with that as well?

----

I think can't do anything right now, as the Measurement Protocol API is in alpha and the other option is using Firebase.

Just going to be patient and wait for things to mature and then work from there. Clearly this was a rush job to pass some legal issues!

**Comment by Eoler** [comment]
I wouldn't stop using v3 (Universal) while it's available (existing dataLayer tracking variables are currently NOT pushed to GA4).

Legality of GA3/4 will be determined by EU DPO's, not wishful thinking of Google or anyone else.

GTag.js is great library and available in this plugin from version (edge?) 1.3.0.

So - _stay calm and track with both accounts_...

**Comment by ayumi-cloud** [comment]
https://developers.google.com/analytics/devguides/config/admin/v1

**Comment by RomainMazB** [comment]
Hi everyone here!
I've just spent 2 hours (+ 3 hours few days ago) struggling to understand WHY THIS F*** SERVICE ACCOUNT CAN'T ACCESS MY GA4 VIEW!

And then I found this issue. I can confirm the plugin is no more compatible with the new GA4 API.

But still, We could create GA3+GA4 Account in Google Analytics, using the "Advance configuration" in the GA Account creation:
![image](https://user-images.githubusercontent.com/53976837/99308741-af681880-2858-11eb-8ff0-f75d8d65e16c.png)
*Sorry, didn't found how to force the english version on GA*

Using this option create two account: GA3+GA4
![image](https://user-images.githubusercontent.com/53976837/99308826-d1fa3180-2858-11eb-8d15-427e2f2b7f70.png)
I'll go this way, and delete the GA3 account when the plugin will be compatible!

Thanks @ayumi-cloud for your investigation and @WoMec for your issue report!

**Comment by ayumi-cloud** [comment]
https://issuetracker.google.com/issues/173422062

**Comment by grafxflow** [comment]
Would it be better to create a new plugin rather than change the current one?

**Comment by ayumi-cloud** [comment]
> Would it be better to create a new plugin rather than change the current one?

My two cents is like this:

1. This plugin is used by other plugins, if we created a new plugin, you would break other plugins and need to get the authors of those plugins to change their code.

2. People seem to hold on to GA3, from day one I couldn't understand how Google got away with breaking GDPR laws. A lot of changes in the link provided by Eoler shows Google has changed their code to pass GDPR. Though GDPR is a complete JOKE! As no company that has broken GDPR has had a large fine yet! In a perfect world, Google should of been fined up to 2% of it annual revenue. Because of this large amount of money, I would be surprised Google would keep GA3 going that long and close it down asap.

3. Changing from GA3 to GA4 was pretty easy, just register a new account property under the same account and change the embed code on your websites.

e.g.

https://github.com/rainlab/googleanalytics-plugin/blob/d013ee4841afc9a29baea2fd910ae0c1feb13ee8/components/tracker/default.htm#L1-L16

Would turn into this:

```php
<script async src="https://www.googletagmanager.com/gtag/js?id={{ __SELF__.trackingId }}"></script>
<script>
window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', '{{ __SELF__.trackingId }}');
</script>
```

Plus you could also add another component to support AMP-HTML code, which would look something like this:

```html
<script async custom-element="amp-analytics" src="https://cdn.ampproject.org/v0/amp-analytics-0.1.js"></script>
<amp-analytics type="gtag" data-credentials="include">
<script type="application/json">
{
  "vars" : {
    "gtag_id": "{{ __SELF__.trackingId }}",
    "config" : {
      "{{ __SELF__.trackingId }}": { "groups": "default" }
    }
  }
}
</script>
</amp-analytics>
```

**Comment by ayumi-cloud** [comment]
> Would it be better to create a new plugin rather than change the current one?

My two cents is like this:

1. This plugin is used by other plugins, if we created a new plugin, you would break other plugins and need to get the authors of those plugins to change their code.

2. People seem to hold on to GA3, from day one I couldn't understand how Google got away with breaking GDPR laws. A lot of changes in the link provided by Eoler shows Google has changed their code to pass GDPR. Though GDPR is a complete JOKE! As no company that has broken GDPR has had a large fine yet! In a perfect world, Google should of been fined up to 2% of it annual revenue. Because of this large amount of money, I would be surprised Google would keep GA3 going that long and close it down asap.

3. Changing from GA3 to GA4 was pretty easy, just register a new account property under the same account and change the embed code on your websites.

e.g.

https://github.com/rainlab/googleanalytics-plugin/blob/d013ee4841afc9a29baea2fd910ae0c1feb13ee8/components/tracker/default.htm#L1-L16

Would turn into this:

```php
<script async src="https://www.googletagmanager.com/gtag/js?id={{ __SELF__.trackingId }}"></script>
<script>
window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', '{{ __SELF__.trackingId }}');
</script>
```

As the `domainName`, `forceSSL` and `anonymizeIp` are automatic now and so can be removed.

Plus you could also add another component to support AMP-HTML code, which would look something like this:

```html
<script async custom-element="amp-analytics" src="https://cdn.ampproject.org/v0/amp-analytics-0.1.js"></script>
<amp-analytics type="gtag" data-credentials="include">
<script type="application/json">
{
  "vars" : {
    "gtag_id": "{{ __SELF__.trackingId }}",
    "config" : {
      "{{ __SELF__.trackingId }}": { "groups": "default" }
    }
  }
}
</script>
</amp-analytics>
```

[edit] I would +1 for the idea of creating a `ga4` branch to this plugin instead.

**Comment by LukeTowers** [comment]
@grafxflow make a PR to this one and we'll review it.

**Comment by ayumi-cloud** [comment]
Might as well wait, I already spoken to a Googler and they have put in a request to build a php sdk for the api now.

**Comment by ayumi-cloud** [comment]
Not really feeling this plugin, just going to hard fork and re-write this plugin from the ground up.

Going to add a lot more features and remove some questionable code.

People are free to add any pr's they like to it.

Repo found here: https://github.com/ayumi-cloud/ga4

**Comment by ayumi-cloud** [comment]
Been running some tests with GA4 in October.

The data gathering in many browsers is really poor!

For example, the latest Firefox, Safari, Brave etc. block the GA4 tracker by default, see screenshot:

![image](https://user-images.githubusercontent.com/57409060/99905284-d646a480-2cc7-11eb-9a45-4658633944f3.png)

Above: Firefox (by default) no settings changed, blocks the tracker trying to send the data via a POST 204 status code.

To make this plugin work, you would need to send the data via a proxy and not from an internal filename called: `analytics.js`

Writing this comment as a note, for the authors to consider.

**Comment by ayumi-cloud** [comment]
GA Reporting API v4 PHP SDK here: 

https://developers.google.com/analytics/devguides/reporting/core/v4/quickstart/web-php

But I think this plugin needs: **Google Analytics Admin API (GA4)** Found here: 

https://developers.google.com/analytics/devguides/config/admin/v1#client_libraries

p.s. If it comes out - will try my best to write an update comment.

**Comment by ayumi-cloud** [comment]
I'm messing about with this plugin here: https://github.com/ayumi-cloud/ga4

(but I don't want to make any guarantees about completely things).

**Comment by SaifurRahmanMohsin** [comment]
I can't wait for Google to roll out a PHP SDK for this. They already have an apiclient library and I used that to make a [fork of this plugin](https://github.com/mohsin/googleanalytics-plugin). Might even extend it to provide Mobile features since I work mainly with Mobile on OctoberCMS (even put up a series of plugins related to it).

**Comment by LukeTowers** [comment]
@SaifurRahmanMohsin feel free to make a PR

**Comment by daftspunk** [comment]
Added a note about the legacy accounts. Can upgrade the plugin later

**Comment by chocolata** [comment]
Hi guys, what's the status on this? Is an update to GA4 on the horizon? Or is a workaround for the above problems available yet?

**Comment by daftspunk** [comment]
> Is an update to GA4 on the horizon?

Yes, once UA is properly shut down, then the path becomes clear since this plugin will not need to provide support with both.
